dream.py

```diff
@@ -144,40 +208,193 @@ def generate_data(self, n_events):
+            
+            verbose: Boolean
+                Set to True for a short analysis of generated data to be written to terminal
         """
-        self.hist = sc.histogram(self.dspacing, dspacing.coords[Dim.DSpacing])
-        return self.hist
+    
+        if wavelength_center - wavelength_width/2.0 < 0:
+            raise ValueError("Provided wavelength band with negative wavelengths.")
+    
+        # get pixel theta:
+        positions = np.array(dream.d.coords[Dim.Position].values)
+        n_positions = (len(positions))
+        z_dir = np.zeros((n_positions, 3))
+        z_dir[:,2] = 1.0
+        position_lengths = np.sqrt(np.sum(positions**2, axis=1))
+        position_zdir_dots = np.sum(positions*z_dir, axis=1)
+        tthetas = 180/np.pi*np.array(np.arccos(position_zdir_dots/position_lengths))
+        thetas = 0.5*tthetas
+        
+        # Find allowed bragg peaks for each pixel and find total strength in order to normalize
+        allowed = np.full((n_positions, len(self.sample_bragg_peaks)), False)
+        pixel_strengths = np.zeros(n_positions)
+        count = 0
+        for bragg_peak in self.sample_bragg_peaks:
+            wavelength = 2.0*bragg_peak.d_spacing*np.sin(thetas)
+            allowed[:,count] = np.logical_and(wavelength < wavelength_center + wavelength_width/2.0,
+                                              wavelength > wavelength_center - wavelength_width/2.0)
+            pixel_strengths[allowed[:,count]] += bragg_peak.strength
+            count += 1
+            
+        # Now we have pixel_strength as indicator for relative intensity in each peak (assuming relatively narrow peaks)
+        total_strength = np.sum(pixel_strengths)
+        events_per_strength = n_events / total_strength
 
-    #def data_reduction(self):
-        # "DiffractionFocussing" == sum? (not available yet)
-        # focussed = sc.sum(hist, Dim.Position)
+        # Calculate distance from source to each pixel when scattered in sample
+        source_to_pixel = self.source_sample_dist + position_lengths
+        
+        # Prepare progress bar and timing of most time consuming loop
+        progress_fraction = int(n_positions/10)
+        progress_count = 0
+        start_time = time.time()
+        
+        print("Generating sparse data for each pixel")
+        # For each pixel we check all allowed bragg peaks and generate a poisson distributed number of events from each
+        for pixel_id in range(n_positions):
+        
+            if progress_bar and pixel_id % progress_fraction == 0:
+                print(str(progress_count*10) + "%")
+                progress_count += 1
+            
+            allowed_bragg_peaks = allowed[pixel_id,:]
+            wavelengths = []
+            for bragg_peak_id in np.arange(len(allowed_bragg_peaks))[allowed_bragg_peaks]:
+                returned_wavelengths = 2.0*np.sin(thetas[pixel_id])*self.sample_bragg_peaks[bragg_peak_id].generate_events(events_per_strength)
```

Comment:
Is the call to `generate_events` the slow part? Would it help to just use the same random numbers all the name, and simply take a slice of required length, and then scale/offset it?

Reply:
I believe the root of the performance problem is that it is two nested for loops that need to work with small arrays. Due to the input interface for sparse data, a for loop over detector pixels is mandatory. I think the problem is to hold the entire pixel array in memory, as the execution time scales almost only with this parameter and changes slope at a critical length. See quick graph of execution times.
As I understand it, your proposed fix would improve the scaling with n_events, but the problem lies with the scaling of n_pixels.

![IMG_0527](https://user-images.githubusercontent.com/24249970/64096956-42578880-cd63-11e9-90e6-4a7c984129ff.jpg)